### PR TITLE
JIP-16 Fake Data and Setup Database

### PR DESCRIPTION
**Pull request by jpfranci**
## Database Schemas
* To look at the database schemas, look at any file called schema in this PR
* I created 5 collections -- users, posts, upvotes, downvotes, tags
* Upvotes/downvotes contains data of the user and media id (post or comment id) that it belongs to
* Comments is nested inside of posts (Not sure if we prefer it this way or want to just store it in its own collection. If we store in its own collection then we should cache the number of comments in there.)
* For efficiency of querying, I cached usernames in both posts and comments (we will have to either change this or deal with having to update in multiple places)

## MongoDB Setup
* Everytime you start the development server, `mongod` will also be automatically be started (if it wasn't already)

## Fake Data Script
To manually run the script, you can just run `node datagen.js` and pass in any option you want to it. If you do `node datagen.js -h ` you can get a help screen of what options are available.
image

The script will also run when you do `yarn setup`.

### What it does
1) Loads fake data from datasets on Reddit comments, Reddit posts, Metro Vancouver latitude/lon data (I manually got that from dragging our google map haha), and tags (some degen ones)
2) Randomizes number of upvotes and downvotes on each post/comment matching it on the existing score on every run
3) Randomizes location of each post
4) Randomizes usernames and emails on every run
5) Wipes database (if you set reset option to true)
6) Inserts all documents into the db (this can take up to 5 min because there's like up to 200k documents)

### CODE REVIEW

File: packages/server/setup/datagen.js [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+      numDownvotes = chance.integer({
+        min: Math.abs(scoreToUse),
+        max: Math.abs(scoreToUse) * 2,
+      });
+      numUpvotes = numDownvotes + scoreToUse;
+    }
+
+    const voters = [...this.users];
+    this.generateVoters(voters, id, numUpvotes, this.upvotes);
+    this.generateVoters(voters, id, numDownvotes, this.downvotes);
```
Comment (srhsrhsrhsrh) [inline packages/server/setup/datagen.js:123]: DEMOCRACY (✧ω✧)

File: packages/server/setup/mock-data/tags_bank.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  "bets",
+  "yolo",
+  "dd",
+  "tendies",
+  "diamond hands",
+  "stonks",
+  "funds",
+  "gains",
+  "loss"
+]
```
Comment (srhsrhsrhsrh) [inline packages/server/setup/mock-data/tags_bank.json:11]: very nice

File: packages/server/src/data/schemas/post-schema.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  body: String,
+  date: Date,
+  numUpvotes: Number,
+  numDownvotes: Number,
+  location: LocationSchema,
+  comments: [CommentSchema],
+  userId: ObjectId,
+  username: String,
+  tag: String,
+});
```
Comment (srhsrhsrhsrh) [inline packages/server/src/data/schemas/post-schema.js:23]: just checking... no more anonymous posting?

**Review by srhsrhsrhsrh — APPROVED** [review]

File: packages/server/src/data/schemas/post-schema.js (hunk 3 above)
Comment (jpfranci) [reply packages/server/src/data/schemas/post-schema.js:23]: If we want to do anonymous, I guess we can just make username/userId undefined. Basically anonymous posts don't count towards your rep.

**Pull request merged**